### PR TITLE
[0.2] Unsubscribe

### PR DESCRIPTION
**Pull request by dasdo**
add method to unsubscribe from emails

[POST] /v1/users/<usersID|0>/unsubscribe
[Params] <notification_types> array of int 

// Unsubscribe list 1
notification_types[0] = 1
// Unsubscribe all
notification_types[1] = -1

### CODE REVIEW

**Review by rwhite27 — APPROVED** [review]

File: routes/api.php [hunk 1]
```diff
@@ -27,6 +27,7 @@
     Route::crud('/apps-keys')->controller('AppsKeysController'),
     Route::post('/apps-keys/regenerate')->controller('AppsKeysController')->action('regenerateKeys'),
     Route::crud('/users')->notVia('post'),
+    Route::post('/users/{id}/unsubscribe')->controller('UsersController')->action('unsubscribe'),
```
Comment (kaioken) [inline routes/api.php:31]: @dasdo @SparoHawk if im unsubscribing shouldnt it be a delete?

File: src/Api/Controllers/UsersController.php [hunk 2]
```diff
@@ -244,4 +246,40 @@ public function changeAppUserActiveStatus(int $id, int $appsId) : Response
...
+            $id = $this->userData->getId();
+        }
+
+        $user = $this->model->findFirstOrFail([
+            'id = ?0 AND is_deleted = 0',
+            'bind' => [$id],
+        ]);
+        $unsubscribe = [];
+        foreach ($request['notification_types'] as $typeId) {
+            $notificationType = NotificationType::findFirst($typeId);
```
Comment (kaioken) [inline src/Api/Controllers/UsersController.php]: @dasdo use findfirstorFail , no?

File: src/Api/Controllers/UsersController.php [hunk 3]
```diff
@@ -244,4 +246,40 @@ public function changeAppUserActiveStatus(int $id, int $appsId) : Response
...
+            $systemModulesId = $notificationType ? $notificationType->system_modules_id : -1;
+            $unsubscribe[] = Notifications::unsubscribe($user, (int) $typeId, (int) $systemModulesId);
```
Comment (kaioken) [inline src/Api/Controllers/UsersController.php]: @dasdo its good for now ? but shouldnt this metho of unbuscribe be on the users end?

Im a user who is unsubscribing from the type X?

so it should be $user->unsubcribe(NotificationType)?

File: src/Models/Notifications.php [hunk 4]
```diff
@@ -84,4 +84,26 @@ public static function totalUnRead(Users $user) : int
             ]
         ]);
     }
+
+    /**
+     * Verify that the user is unsubscribed
+     * @return bool
+     */
+    public function isUnsubscribe(?Users $user) : bool
```
Comment (kaioken) [inline src/Models/Notifications.php]: @dasdo I thinks this should be on the users end

**Review by kaioken — CHANGES_REQUESTED** [review]

File: routes/api.php (hunk 1 above)
Comment (dasdo) [reply routes/api.php:31]: Information is being added to the database, so it's post

File: src/Api/Controllers/UsersController.php (hunk 2 above)
Comment (dasdo) [reply src/Api/Controllers/UsersController.php]: No, we can receive -1 which is not a list.

**Review by kaioken — APPROVED** [review]

**Review by rwhite27 — APPROVED** [review]

**Pull request merged**